examples/calculator/lib/calc_expression.dart

```diff
@@ -0,0 +1,329 @@
+// Copyright 2016 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+/// A token that composes an expression. There are several kinds of tokens
+/// that represent arithmetic operation symbols, numbers and pieces of numbers.
+/// We need to represent pieces of numbers because the user may have only
+/// entered a partial expression so far.
+class ExpressionToken {
+  String stringRep;
+
+  ExpressionToken(this.stringRep);
+
+  String toString() {
+    return stringRep;
+  }
+}
+
+// A token that represents an integer.
+class IntToken extends ExpressionToken {
+  int number;
+
+  IntToken(String stringRep) : super(stringRep) {
+    number = int.parse(stringRep);
+  }
+}
+
+// A token that represents a floating point number.
+class FloatToken extends ExpressionToken {
+  double number;
+
+  FloatToken(String stringRep) : super(stringRep) {
+    var toParse = stringRep;
+    if (toParse.startsWith(".")) {
+      toParse = "0" + toParse;
+    }
+    if (toParse.endsWith(".")) {
+      toParse = toParse + "0";
+    }
+    number = double.parse(toParse);
+  }
+}
+
+// A token that represents a number that is the result of a computation.
+class ResultToken extends ExpressionToken {
+  num number;
+
+  ResultToken(num number)
+      : super("$number"),
+        this.number = number;
+}
+
+// A token that represents the unary minus prefix.
+class LeadingNegToken extends ExpressionToken {
+  LeadingNegToken() : super("-");
+}
+
+enum Operation { Addition, Subtraction, Multiplication, Division }
+
+// A token that represents an arithmetic operation symbol.
+class OperationToken extends ExpressionToken {
+  Operation operation;
+
+  OperationToken(Operation operation)
+      : super(opString(operation)),
+        this.operation = operation;
+
+  static String opString(Operation operation) {
+    switch (operation) {
+      case Operation.Addition:
+        return " + ";
+      case Operation.Subtraction:
+        return " - ";
+      case Operation.Multiplication:
+        return "  \u00D7  ";
+      case Operation.Division:
+        return "  \u00F7  ";
+    }
+  }
+}
+
+// As the user taps different keys the current expression can be in one
+// of several states.
```

Comment:
If you use `///` rather than `//`, these comments will be automatically picked up by the `dartdoc` tool as documentation.